alphagov/paas-admin
# [#166028126] MS SSO flow

What
----

This a very large commit, which touches a number of things. The three major
changes are:

  1. New environment variables for Microsoft Active Directory
     credentials, and application domain name
  2. Implementation of handling an Open ID Connect authorization code grant handshake
  3. UI flow for guiding users through the uplift.

For GDS users of the Google SSO provider, we were able to safely assume that
the username we held for them would be the same as their Google account email
address. However, for Microsoft users we know this isn't the case.

For the Microsoft SSO provder, we map the `oid` claim (which contains
Microsoft's identifier for the user) to our `userName` field in UAA. This means
that when we change a user's origin from `uaa` to `microsoft`, we
simultaneously need to change their username to match the value of the claim.
Otherwise, UAA will be unable to line up the two fields and identify the user,
and thus the user cannot log in.

In order to get the `oid` claim for a user, we need to have them log in with
Microsoft, and then request some information about the identity of the person.
For that, we use Open ID Connect flow to request an identity token. We
subequently pull the `oid` claim from the token, and update the `userName` and
`origin` in UAA in one API call.

In the future, we might choose to implement this OIDC flow for non-GDS Google
users. In that event, we will need to modify this flow and add some
configurable claim mapping.

How to review
-------------
1. Deploy this [`paas-cf` PR](https://github.com/alphagov/paas-cf/pull/1961) which contains this change, and the changes required in `paas-uaa-customized-boshrelease` and `paas-cf`
2. In your dev env, invite a new user
3. Log in as the new user
4. Follow the MS SSO uplift flow 
5. Log out
6. Log in with MS SSO, and see you're logged in to the same user account as before

Who can review
---------------
Anyone but @AP-Hunt and @mogds
